src/scale_impls/decode.rs

```diff
@@ -86,69 +92,80 @@ impl scale_decode::DecodeAsFields for Composite<()> {
         types: &'resolver R,
     ) -> Result<Self, scale_decode::Error> {
         // Build a Composite type to pass to a one-off visitor:
-        let mut composite =
-            scale_decode::visitor::types::Composite::new(input, fields, types, false);
+        let mut composite = scale_decode::visitor::types::Composite::new(
+            core::iter::empty(),
+            input,
+            fields,
+            types,
+            false,
+        );
         // Decode into a Composite value from this:
-        let val = visit_composite::<R, (), DefaultMapper>(&mut composite);
+        let val = visit_composite::<R, (), EmptyContext>(&mut composite);
         // Consume remaining bytes and update input cursor:
         composite.skip_decoding()?;
         *input = composite.bytes_from_undecoded();
         val.map_err(From::from).map(|v| v.map_context(|_| {}))
     }
 }
 
-/// A [`scale_decode::Visitor`] implementation for decoding into [`Value`]s.
-pub struct DecodeValueVisitor<R: TypeResolver, T, F> {
-    resolver: PhantomData<(R, T, F)>,
-}
-impl<R: TypeResolver, T, F> Default for DecodeValueVisitor<R, T, F> {
-    fn default() -> Self {
-        Self::new()
-    }
-}
-
-impl<R: TypeResolver, T, F> DecodeValueVisitor<R, T, F> {
-    pub fn new() -> Self {
-        DecodeValueVisitor { resolver: PhantomData }
-    }
-}
-
 impl scale_decode::IntoVisitor for Value<()> {
     // Note: the DefaultMapper just removes all type ids here.
     type AnyVisitor<R: scale_decode::TypeResolver> =
-        scale_decode::visitor::VisitorWithCrateError<DecodeValueVisitor<R, (), DefaultMapper>>;
+        scale_decode::visitor::VisitorWithCrateError<DecodeValueVisitor<R, (), EmptyContext>>;
 
     fn into_visitor<R: scale_decode::TypeResolver>() -> Self::AnyVisitor<R> {
         scale_decode::visitor::VisitorWithCrateError(DecodeValueVisitor::new())
     }
 }
 
-pub trait TypeIdMapper<From, To> {
-    fn map(from: &From) -> To;
+/// We can use [`DecodeValueVisitor`] to decode values, but have two cases to handle:
+///
+/// - We need to be able to decode into [`Value<()>`] for our [`scale_decode::IntoVisitor`]
+///   implementation above (because that trait is agnostic over which visitor is used, and thus
+///   can't have any knowledge on the TypeId type).
+/// - We need to be able to decode into [`Value<TypeId>`] via the [`decode_value_as_type`] fn
+///   above.
+///
+/// This trait basically allows us to handle each case by having a function that is given a
+/// `TypeId` and decides whether to hand back `()` or the `TypeId`.
+trait ContextFromTypeId<TypeId, Output> {
```

Comment:
Do we need to implement to this for several output types? My question is really why Output isn't an associated type?